### PR TITLE
Зайцев Артем. Задача 1. Вариант 4. Минимальный элемент вектора.

### PR DESCRIPTION
**Pull request by aortemon**
### Последовательная реализация
Принять, что первый элемент вектора, является текущим минимальным найденным. 
Сравнить текущий минимальный элемент с очередным непроверенным элементом вектора, начиная со второго.
При достижении конца вектора текущий минимальный элемент будет минимальным элементом во всем векторе.
### Параллельная реализация
Разбить вектор на подвекторы длины, равной соотношению длины исходного вектора к количеству выделенных потоков.
В каждом потоке запустить последовательную реализацию алгоритма для соответствующего подвектора.
С помощью `boost::mpi::reduce()` найти наименьший элемент среди минимумов, полученных в каждом потоке.

### CODE REVIEW

**Comment by aortemon** [comment]
I've closed my previous PR #189 due to some issues with builds and tests.

@Sean0450 I've added tests for singleton vectors as you requested in #189.

@kmichaelk Thank you for your advice about `std::copy()` given in previous PR. I will use it for upcoming jobs, but not this one, so as not to burden the repo's workers with another commit.

**Review by Sean0450 — APPROVED** [review]

**Review by kmichaelk — APPROVED** [review]

File: tasks/mpi/zaitsev_a_min_of_vector_elements/src/ops_mpi.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+#include "mpi/zaitsev_a_min_of_vector_elements/include/ops_mpi.hpp"
+
+#include <algorithm>
+#include <functional>
+#include <random>
+#include <string>
+#include <thread>
+#include <vector>
+
+using namespace std::chrono_literals;
```
Comment (allnes) [inline tasks/mpi/zaitsev_a_min_of_vector_elements/src/ops_mpi.cpp:11]: please remove it

File: tasks/seq/zaitsev_a_min_of_vector_elements/src/ops_seq.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
+// Copyright 2024 Nesterov Alexander
+#include "seq/zaitsev_a_min_of_vector_elements/include/ops_seq.hpp"
+
+#include <algorithm>
+#include <thread>
+
+using namespace std::chrono_literals;
```
Comment (allnes) [inline tasks/seq/zaitsev_a_min_of_vector_elements/src/ops_seq.cpp:7]: please remove it

File: tasks/mpi/zaitsev_a_min_of_vector_elements/src/ops_mpi.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+std::vector<int> zaitsev_a_min_of_vector_elements_mpi::getRandomVector(int sz, int minRangeValue, int maxRangeValue) {
```
Comment (allnes) [inline tasks/mpi/zaitsev_a_min_of_vector_elements/src/ops_mpi.cpp:13]: please move it to tests files

**Review by allnes — CHANGES_REQUESTED** [review]

**Review by allnes — APPROVED** [review]

File: tasks/mpi/zaitsev_a_min_of_vector_elements/include/ops_mpi.hpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+#include <numeric>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "core/task/include/task.hpp"
+
+namespace zaitsev_a_min_of_vector_elements_mpi {
+
+std::vector<int> getRandomVector(int sz, int minRangeValue, int maxRangeValue);
```
Comment (aobolensk) [inline tasks/mpi/zaitsev_a_min_of_vector_elements/include/ops_mpi.hpp:18]: Please, move test function to the test module

File: tasks/mpi/zaitsev_a_min_of_vector_elements/perf_tests/main.cpp [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+#include "mpi/zaitsev_a_min_of_vector_elements/include/ops_mpi.hpp"
+
+TEST(zaitsev_a_min_of_vector_elements_mpi, test_pipeline_run) {
+  const int extrema = -1;
+  const int minRangeValue = 0;
+  const int maxRangeValue = 1000;
+
+  boost::mpi::communicator world;
+  std::vector<int> global_vec;
+  std::vector<int32_t> global_min(1, maxRangeValue + 1);
```
Comment (aobolensk) [inline tasks/mpi/zaitsev_a_min_of_vector_elements/perf_tests/main.cpp:17]: Please, add test for negative values as well

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: tasks/mpi/zaitsev_a_min_of_vector_elements/include/ops_mpi.hpp (hunk 4 above)
Comment (aortemon) [reply tasks/mpi/zaitsev_a_min_of_vector_elements/include/ops_mpi.hpp:18]: Moved

File: tasks/mpi/zaitsev_a_min_of_vector_elements/perf_tests/main.cpp (hunk 5 above)
Comment (aortemon) [reply tasks/mpi/zaitsev_a_min_of_vector_elements/perf_tests/main.cpp:30]: Tests have been added, already existed ones have been corrected for the case you specified

**Review by aobolensk — APPROVED** [review]
